sql/2024/04/inaccurate-sizes-attribute-impact.sql

```diff
@@ -0,0 +1,90 @@
+# HTTP Archive query to measure impact of inaccurate sizes attributes per <img> for WordPress sites.
+#
+# WPP Research, Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+CREATE TEMPORARY FUNCTION
+  getImgSizesAccuracy(custom_metrics STRING)
+  RETURNS ARRAY<STRUCT<sizesAbsoluteError INT64,
+  sizesRelativeError FLOAT64,
+  idealSizesSelectedResourceEstimatedPixels INT64,
+  actualSizesEstimatedWastedLoadedPixels INT64,
+  idealSizesSelectedResourceEstimatedBytes FLOAT64,
+  actualSizesEstimatedWastedLoadedBytes FLOAT64>>
+  LANGUAGE js AS '''
+try {
+  const $ = JSON.parse(custom_metrics);
+  let responsiveImages = JSON.parse($.responsive_images);
+  responsiveImages = responsiveImages['responsive-images'];
+
+  return responsiveImages.map((imgData) => {
+    return {
+      imgData.sizesAbsoluteError,
+      imgData.sizesRelativeError,
+      imgData.idealSizesSelectedResourceEstimatedPixels
+      imgData.actualSizesEstimatedWastedLoadedPixels,
+      imgData.idealSizesSelectedResourceEstimatedBytes
+      imgData.actualSizesEstimatedWastedLoadedBytes,
```

Comment:
I guess I was assuming each quartile would be relative to each other, so we could easily calculate the % based on these numbers if we decided we wanted to. Perhaps I'm misunderstanding how each of these `APPROX_QUANTILES` values relate to each other.

I'm happy to add back relative values to the query, but am unsure how to best do so without being able to do some trial and error on the query itself. Would it be something like this?

```sql
CREATE TEMPORARY FUNCTION GET_IMG_SIZES_ACCURACY(custom_metrics STRING) RETURNS
  ARRAY<STRUCT<hasSrcset BOOL,
  hasSizes BOOL,
  sizesAbsoluteError INT64,
  sizesRelativeError FLOAT64,
  idealSizesSelectedResourceEstimatedPixels INT64,
  actualSizesEstimatedWastedLoadedPixels INT64,
  relativeSizesEstimatedWastedLoadedPixels FLOAT64,
  idealSizesSelectedResourceEstimatedBytes FLOAT64,
  actualSizesEstimatedWastedLoadedBytes FLOAT64,
  relativeSizesEstimatedWastedLoadedBytes FLOAT64>>
AS (
  ARRAY(
    SELECT AS STRUCT
      CAST(JSON_EXTRACT_SCALAR(image, '$.hasSrcset') AS BOOL) AS hasSrcset,
      CAST(JSON_EXTRACT_SCALAR(image, '$.hasSizes') AS BOOL) AS hasSizes,
      CAST(JSON_EXTRACT_SCALAR(image, '$.sizesAbsoluteError') AS INT64) AS sizesAbsoluteError,
      CAST(JSON_EXTRACT_SCALAR(image, '$.sizesRelativeError') AS FLOAT64) AS sizesRelativeError,
      CAST(JSON_EXTRACT_SCALAR(image, '$.idealSizesSelectedResourceEstimatedPixels') AS INT64) AS idealSizesSelectedResourceEstimatedPixels,
      CAST(JSON_EXTRACT_SCALAR(image, '$.actualSizesEstimatedWastedLoadedPixels') AS INT64) AS actualSizesEstimatedWastedLoadedPixels,
      SAFE_DIVIDE(
        CAST(JSON_EXTRACT_SCALAR(image, '$.idealSizesSelectedResourceEstimatedPixels') AS INT64),
        CAST(JSON_EXTRACT_SCALAR(image, '$.actualSizesEstimatedWastedLoadedPixels') AS INT64)
      ) AS relativeSizesEstimatedWastedLoadedPixels,
      CAST(JSON_EXTRACT_SCALAR(image, '$.idealSizesSelectedResourceEstimatedBytes') AS FLOAT64) AS idealSizesSelectedResourceEstimatedBytes,
      CAST(JSON_EXTRACT_SCALAR(image, '$.actualSizesEstimatedWastedLoadedBytes') AS FLOAT64) AS actualSizesEstimatedWastedLoadedBytes,
      SAFE_DIVIDE(
        CAST(JSON_EXTRACT_SCALAR(image, '$.idealSizesSelectedResourceEstimatedBytes') AS FLOAT64),
        CAST(JSON_EXTRACT_SCALAR(image, '$.actualSizesEstimatedWastedLoadedBytes') AS FLOAT64)
      ) AS relativeSizesEstimatedWastedLoadedBytes,
    FROM
      UNNEST(JSON_EXTRACT_ARRAY(custom_metrics, '$.responsive_images.responsive-images')) AS image
  )
);
```

Reply:
To make ☝🏻 easier to test, I went ahead and pushed [1054716](https://github.com/GoogleChromeLabs/wpp-research/pull/108/commits/1054716da6f42700f7dfbbe37a4df4cb0c5d6cd8) as a first attempt at this.

Reply:
This looks exactly right. Let me try to run the query to verify.

Reply:
> I guess I was assuming each quantiles would be relative to each other, so we could easily calculate the % based on these numbers if we decided we wanted to. Perhaps I'm misunderstanding how each of these `APPROX_QUANTILES` values relate to each other.

The quantiles are only based on the values, but not relatively. So if you have 10 values, with 8 values being `1` and 2 values being `10`, almost all quantiles would show the value `1`. Of the percentiles displayed here only the 90th percentile would not show `1` but `10`.

So my feedback to include relative values here is not related to the quantiles, but about that the absolute sizes error is very much dependent on the size of the images. From an error perspective though, we shouldn't give larger images more impact in our data assessment.

For example, if for an ideal image of 2000px or 1MB an image of 4000px or 2MB is loaded, absolutely speaking, this is a lot worse than for an ideal image of 200px or 100kB loading an image of 400px or 200kB. But relatively, it's the same: The actually loaded image is 100% larger (or 200% as large) compared to the ideal image.

Reply:
I see what you're saying. Showing relative waste gives us a different view of the effect that incorrect sizes attributes have. I'd anticipate that any additional constraint on max sizes values will likely have a greater impact on larger images, since those are the ones that are more likely to be shown when a smaller source is selected, but having both values will make it easier to confirm.

Reply:
After further review, I have one point of feedback regarding:
> ```sql
>       SAFE_DIVIDE(
>         CAST(JSON_EXTRACT_SCALAR(image, '$.idealSizesSelectedResourceEstimatedPixels') AS INT64),
>         CAST(JSON_EXTRACT_SCALAR(image, '$.actualSizesEstimatedWastedLoadedPixels') AS INT64)
>       ) AS relativeSizesEstimatedWastedLoadedPixels,
>       ...
>       SAFE_DIVIDE(
>         CAST(JSON_EXTRACT_SCALAR(image, '$.idealSizesSelectedResourceEstimatedBytes') AS FLOAT64),
>         CAST(JSON_EXTRACT_SCALAR(image, '$.actualSizesEstimatedWastedLoadedBytes') AS FLOAT64)
>       ) AS relativeSizesEstimatedWastedLoadedBytes,
> ```

I'm unsure the order of this division is right. I don't necessarily have the answer myself, but I'm still trying to decipher what the two values respectively are. To look at some real data, I ran a the "sub query" part of this (i.e. without aggregating into percentiles) for a specific URL, in this case `https://wordpress.org/`. I've put the query results into [this spreadsheet](https://docs.google.com/spreadsheets/d/1RKfXU14r9KRoQPzb8RN_bhLiBM2M1p8B6GkG10oOkVQ/edit) for us to look at. This basically has the data for every single image on that page.

What I see is that the `actualSizes...` value is many times `0`. Maybe we should verify at the actual page for whether that makes sense based on our understanding.

In my understanding so far, the `actualSizes...` is the amount of pixels/bytes that is wasted, and the `idealSizes...` is the amount of pixels/bytes based on the ideal available srcset. If that's the case, it would mean the sum of the two values is the amount of pixels/bytes actually loaded. Based on that, to get the relevant wasted loaded pixels/bytes, we need to make the division the other way around (`actualSizes... / idealSizes...`). If the loaded image is a little too large, it would give us small percentages for example, while if the image is e.g. twice as large, it would give us 100%, or three times as large would give us 200%.

Reply:
My understanding is that in both the cases of estimated wasted pixels and wasted bytes, `idealSizesSelectedResource...` represents the resource (from `srcset`) that would have been selected if the `sizes` value was accurate. Similarly, `actualSizesEstimatedWasted...` represents the difference between the ideal size and the size of the selected source. So the places where you're seeing 0 reported for `actualSizesEstimatedWasted...` means that even though the `sizes` attribute was inaccurate, there wasn't a better source available in the `srcset`.

That said, I think you're probably right that we would want to divide the estimated wasted bytes by the ideal selected to get a relative value of the wasted value.